src/array.c

```diff
@@ -710,58 +709,85 @@ STATIC_INLINE void jl_array_grow_at_beg(jl_array_t *a, size_t idx, size_t inc,
         size_t oldoffsnb = a->offset * elsz;
         size_t nb1 = idx * elsz;
         if (inc > (a->maxsize - n) / 2 - (a->maxsize - n) / 20) {
+            // not enough room for requested growth from end of array
             size_t newlen = a->maxsize == 0 ? inc * 2 : a->maxsize * 2;
             while (n + 2 * inc > newlen - a->offset)
                 newlen *= 2;
             newlen = limit_overallocation(a, n, newlen, 2 * inc);
             size_t newoffset = (newlen - newnrows) / 2;
-            if (!array_resize_buffer(a, newlen))
+            if (!array_resize_buffer(a, newlen)) {
                 data = (char*)a->data + oldoffsnb;
-            newdata = (char*)a->data + newoffset * elsz;
-            if (isbitsunion) {
-                memmove(newdata + newnrows * elsz, data + n * elsz, idx);
-                memmove(newdata + newnrows * elsz + idx + inc, data + n * elsz + idx, n - idx);
-                memset(newdata + newnrows * elsz + idx, 0, inc);
+                if (isbitsunion) typetagdata = data + oldttoffset;
             }
+            newdata = (char*)a->data + newoffset * elsz;
+            if (isbitsunion) newtypetagdata = newdata + a->maxsize * elsz + newoffset;
             // We could use memcpy if resizing allocates a new buffer,
             // hopefully it's not a particularly important optimization.
-            if (idx > 0 && newdata < data)
+            if (idx > 0 && newdata < data) {
                 memmove(newdata, data, nb1);
+                if (isbitsunion) {
+                    memmove(newtypetagdata, typetagdata, idx);
+                    memset(newtypetagdata + idx, 0, inc);
+                }
+            }
             memmove(newdata + nbinc + nb1, data + nb1, n * elsz - nb1);
-            if (idx > 0 && newdata > data)
+            if (isbitsunion) memmove(newtypetagdata + idx + inc, typetagdata + idx, n - idx);
+            if (idx > 0 && newdata > data) {
                 memmove(newdata, data, nb1);
+                if (isbitsunion) {
+                    memmove(newtypetagdata, typetagdata, idx);
+                    memset(newtypetagdata + idx, 0, inc);
+                }
+            }
             a->offset = newoffset;
         }
         else {
+            // use extra space between a->nrows & a->maxsize
             a->offset = (a->maxsize - newnrows) / 2;
             newdata = data - oldoffsnb + a->offset * elsz;
-            if (isbitsunion) {
-                memmove(newdata + newnrows * elsz + idx + inc, data + n * elsz + idx, n - idx);
-                memset(newdata + newnrows * elsz + idx, 0, inc);
-            }
+            if (isbitsunion) newtypetagdata = newdata + (a->maxsize - a->offset) * elsz + a->offset;
             // We could use memcpy if resizing allocates a new buffer,
             // hopefully it's not a particularly important optimization.
-            if (idx > 0 && newdata < data)
+            if (idx > 0 && newdata < data) {
```

Comment:
I'm not sure this codepath is reachable; the conditions above seem to prevent the case where `inc` will ever be larger than `a->offset`, yet smaller than `(a->maxsize - n) / 2 - (a->maxsize - n) / 20`.

Just a note as I've been trying to ensure we have tests that hit all these various conditional branches.